### PR TITLE
Implement rule metadata service.

### PR DESCRIPTION
**Pull request by malithie**
### Proposed changes in this pull request

Resolves https://github.com/wso2/product-is/issues/21803

In order to populate the rule UI it should be possible to retrieve all fields and field value options that are needed to configure expressions.
This PR implements the core rule metadata service.

### When should this PR be merged

Merging this PR releases the service component and the feature.
Feature will be bundled to the product over a separate PR. So note that merging this PR will not bundle the feature with the product.

### CODE REVIEW

File: components/rule-mgt/org.wso2.carbon.identity.rule.metadata/pom.xml [hunk 1]
```diff
@@ -0,0 +1,156 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (madurangasiriwardena) [inline components/rule-mgt/org.wso2.carbon.identity.rule.metadata/pom.xml:1]: Add license header

File: components/rule-mgt/org.wso2.carbon.identity.rule.metadata/src/main/java/org/wso2/carbon/identity/rule/metadata/provider/RuleMetadataProvider.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+     * @param flowType     Flow type
+     * @param tenantDomain Tenant domain
+     * @return List of field definitions
+     * @throws RuleMetadataException If an error occurred while getting the metadata
+     */
+    List<FieldDefinition> getExpressionMeta(FlowType flowType, String tenantDomain)
+            throws RuleMetadataException;
+
+}
+
```
Comment (madurangasiriwardena) [inline components/rule-mgt/org.wso2.carbon.identity.rule.metadata/src/main/java/org/wso2/carbon/identity/rule/metadata/provider/RuleMetadataProvider.java:45]: unwanted new line

File: components/rule-mgt/org.wso2.carbon.identity.rule.metadata/src/main/java/org/wso2/carbon/identity/rule/metadata/config/FieldDefinitionConfig.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        return operators;
+    }
+
+    private static Value convertValueObject(Map<String, Object> value, ObjectMapper mapper) {
+
+        Map<String, Object> valueObject =
+                mapper.convertValue(value.get("value"), new TypeReference<Map<String, Object>>() {
+                });
+
+        Value.InputType inputType = Value.InputType.valueOf(valueObject.get("inputType").toString().toUpperCase());
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.metadata/src/main/java/org/wso2/carbon/identity/rule/metadata/config/FieldDefinitionConfig.java:107]: `"inputType"` 
Shall we define these as constants in the class ?

File: components/rule-mgt/org.wso2.carbon.identity.rule.metadata/src/main/java/org/wso2/carbon/identity/rule/metadata/util/RuleMetadataExceptionBuilder.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+        // Client Errors.
+        ERROR_INVALID_FLOW_TYPE("60001", "Invalid flow.", "Provided flow type is invalid or not implemented."),
+
+        // Server Errors.
+        ERROR_WHILE_LOADING_STATIC_RULE_METADATA("65001", "Error while loading static rule metadata.",
+                "Server encountered an error while loading rule metadata from server configs."),
+        ERROR_DUPLICATE_FIELD("65002", "Duplicate field found.",
+                "Field: %s from metadata provider: %s already exists in the rule metadata."),
+        ;
```
Comment (ashanthamara) [inline components/rule-mgt/org.wso2.carbon.identity.rule.metadata/src/main/java/org/wso2/carbon/identity/rule/metadata/util/RuleMetadataExceptionBuilder.java:78]:
```suggestion
        ERROR_DUPLICATE_FIELD("65002", "Duplicate field found.",
                "Field: %s from metadata provider: %s already exists in the rule metadata.");
```

File: components/rule-mgt/org.wso2.carbon.identity.rule.metadata/src/main/java/org/wso2/carbon/identity/rule/metadata/config/FieldDefinitionConfig.java (hunk 3 above)
Comment (malithie) [reply components/rule-mgt/org.wso2.carbon.identity.rule.metadata/src/main/java/org/wso2/carbon/identity/rule/metadata/config/FieldDefinitionConfig.java:107]: These stuff are not reused right ? My view is having the literal is more readable

**Comment by malithie** [comment]
Closing this PR as this is conflicted. Refer to https://github.com/wso2/carbon-identity-framework/pull/6164 instead

**Pull request closed without merging**